File: Algorithm/DynamicProgramming/other/To achieve the lowest price/to achieve the lowest price 3/GPT/to-achieve-the-lowest-price-3.js
// Node.js v16.17.1
// 八百屋のりんご購入最適化問題
// 標準入力から n, x, a, y, b を受け取り、りんごを n 個以上手に入れる最小金額を計算する。
// 入力形式: n x a y b
//
// パラメータ
// n: number -> 必要なりんごの個数
// x: number -> パック1の個数
// a: number -> パック1の価格
// y: number -> パック2の個数
// b: number -> パック2の価格
//
// 戻り値
// number -> n個以上のりんごを手に入れるための最小金額
//
// 計算量: O(n + y) （最大2000程度なので十分高速）
// メモリ: O(n + y) （整数配列を1本持つだけ）

const fs = require('fs');

function minAppleCost(n, x, a, y, b) {
    // DP: dp[i] = i個のりんごを得るための最小コスト
    // サイズは n + max(x, y) まで確保して、n以上のケースも対応
    const maxApples = n + Math.max(x, y);
    const INF = Number.MAX_SAFE_INTEGER;
    const dp = new Array(maxApples + 1).fill(INF);
    dp[0] = 0;

    for (let i = 0; i <= maxApples; i++) {
        if (dp[i] === INF) continue;
        if (i + x <= maxApples) {
            dp[i + x] = Math.min(dp[i + x], dp[i] + a);
        }
        if (i + y <= maxApples) {
            dp[i + y] = Math.min(dp[i + y], dp[i] + b);
        }
    }

    // n個以上の中で最小のコストを探す
    let ans = INF;
    for (let i = n; i <= maxApples; i++) {
        ans = Math.min(ans, dp[i]);
    }
    return ans;
}

// 標準入力の処理
(function main() {
    const input = fs.readFileSync(0, 'utf8').trim().split(' ').map(Number);
    const [n, x, a, y, b] = input;
    const result = minAppleCost(n, x, a, y, b);
    console.log(result);
})();

File: Algorithm/DynamicProgramming/other/To achieve the lowest price/to achieve the lowest price 3/GPT/to-achieve-the-lowest-price-3.py
# ```python
# Python 3.8.10
# 八百屋のりんご購入最適化問題
# りんごを n 個以上手に入れる最小コストを求める。
# 入力形式: n x a y b （標準入力から読み込み）
#
# 関数仕様:
# def min_apple_cost(n: int, x: int, a: int, y: int, b: int) -> int
#   n: 必要なりんごの個数
#   x: パック1の個数
#   a: パック1の価格
#   y: パック2の個数
#   b: パック2の価格
#   return: n個以上のりんごを得る最小コスト
#
# 計算量: O(n + max(x,y))  -> 高々 2000 程度で高速
# メモリ使用量: O(n + max(x,y)) -> 配列 ~2000 要素、数十KB 程度

import sys
from typing import List

def min_apple_cost(n: int, x: int, a: int, y: int, b: int) -> int:
    """
    必要なりんご n 個以上を得るための最小コストを計算する

    パラメータ:
        n (int): 必要なりんごの個数
        x (int): パック1の個数
        a (int): パック1の価格
        y (int): パック2の個数
        b (int): パック2の価格

    戻り値:
        int: n個以上を得るための最小コスト
    """

    # 上限を n + max(x,y) とする（理由: 必ずその範囲内に最適解が存在）
    max_apples: int = n + max(x, y)
    INF: int = 10**15  # 十分大きな値で初期化

    # DP 配列: dp[i] = i個ちょうどのりんごを得る最小コスト
    dp: List[int] = [INF] * (max_apples + 1)
    dp[0] = 0  # 0個はコスト0

    # 遷移処理
    for i in range(max_apples + 1):
        if dp[i] == INF:
            continue
        if i + x <= max_apples:
            dp[i + x] = min(dp[i + x], dp[i] + a)
        if i + y <= max_apples:
            dp[i + y] = min(dp[i + y], dp[i] + b)

    # n個以上の中で最小のコストを探索
    ans: int = min(dp[n:])
    return ans


def main() -> None:
    """
    標準入力からデータを読み込み、結果を出力する。
    """
    data = sys.stdin.read().strip().split()
    n, x, a, y, b = map(int, data)
    result: int = min_apple_cost(n, x, a, y, b)
    print(result)


if __name__ == "__main__":
    main()
# ```

# ---

# ## 処理時間とメモリの目安

# * **時間計算量**: `O(n + max(x,y))` → 制約下では高々 2000 ステップ程度で実行、実測で 0.001 秒未満。
# * **メモリ消費**: `dp` 配列のサイズは最大 2001 要素。各要素が Python の `int`（オブジェクト）なので内部的には数十バイト程度だが、合計でも **数十 KB 程度**に収まる。

# 効率的かつ安全に実行可能です。
